### PR TITLE
feature(volume_restrictions): return Skip in PreFilter

### PR DESCRIPTION
**Pull request by AxeZhan**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
return Skip in volume_restrictions PreFilter when the following condition is satisfied:

- pod doesn't have a volume

#### Which issue(s) this PR fixes:

Part of #114399

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:76]: Ditto for other tests in this file

```suggestion
				t.Errorf("PreFilter: status does not match (-want,+got):\n%s", diff)
```

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:81]: Ditto for other tests in this file

```suggestion
					t.Errorf("Filter: status does not match (-want,+got):\n%s", diff)
```

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:66]: Could you help add field names explicitly on each struct? It's a bit hard to see now.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:78]: make it always running Filter regardless of preFilterStatus so that we can confirm if it works as expected when only Filter is enabled.
Ditto for other tests in this file.

**Review by sanposhiho — COMMENTED** [review]
/ok-to-test

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:121]: Actually, even if an incoming pod has Volumes, we sometimes can return Skip. Please refer to `Filter` func.
https://github.com/kubernetes/kubernetes/blob/3aff1f97bef13f1d40e47fae785994d57c3e5a80/pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go#L175

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:134]: We can `return nil, framework.NewStatus(framework.Success)` here, right? Probably like:

```go
	for i := range pod.Spec.Volumes {
		v := &pod.Spec.Volumes[i]
		// fast path if there is no conflict checking targets.
		if v.GCEPersistentDisk == nil && v.AWSElasticBlockStore == nil && v.RBD == nil && v.ISCSI == nil {
			continue
		}
		return nil, framework.NewStatus(framework.Success)
	}
	return nil, framework.NewStatus(framework.Skip)
```

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:62]: Looking much better now, thanks :)

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go
Comment (kidddddddddddddddddddddd) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:134]: nice catch... It's right, we only care if this `volumesCnt >=1`,we don't care the actual number.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:105]: nit:

```suggestion
			filterStatus := p.(framework.FilterPlugin).Filter(ctx, nil, test.pod, test.nodeInfo)
```

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:180]:
```suggestion
			filterStatus := p.(framework.FilterPlugin).Filter(ctx, nil, test.pod, test.nodeInfo)
```

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:261]:
```suggestion
			filterStatus := p.(framework.FilterPlugin).Filter(ctx, nil, test.pod, test.nodeInfo)
```

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:342]:
```suggestion
			filterStatus := p.(framework.FilterPlugin).Filter(ctx, nil, test.pod, test.nodeInfo)
```

**Review by sanposhiho — COMMENTED** [review]
small nits

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go
Comment (kerthcet) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:122]: This is conflict with https://github.com/kubernetes/kubernetes/pull/114051, which changed the logics a lot in PreFilter to support preemption, let's merge that one first. 
/hold

**Review by sanposhiho — APPROVED** [review]
/lgtm for now.

Let's wait for #114051.

**Comment by kidddddddddddddddddddddd** [comment]
pr is updated after #114051 merged.
/cc @kerthcet

**Comment by sanposhiho** [comment]
/assign

I think I can make time to review it this weekend

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go
Comment (kerthcet) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:184]: make this an util function here? It may looks like `func noneConflictVolume(v1.Volume) bool`

Comment (kidddddddddddddddddddddd) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: updated.

**Review by kerthcet — COMMENTED** [review]
/lgtm
/assign @alculquicondor 

It's code freeze now, but maybe we can merge as early as possible in the following release.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go
Comment (kerthcet) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:162]: em...... struct is enough I think,  no need to use pointer here.

Comment (kidddddddddddddddddddddd) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: Updated, but out of curiosity, can you explain why?

Comment (kerthcet) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: Go is a language with GC, so if the struct is not big one, prefer to use struct to avoid GC for the sake of performance.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go
Comment (kerthcet) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:185]: Suddenly came up with a scenario, if pod with no volumes, and we disable the enableReadWriteOncePod, we should return Skip right? Not nil as success. Also we missed the corresponding test.

Comment (kidddddddddddddddddddddd) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: Nice catch! Added one.

**Comment by kerthcet** [comment]
A strong feeling came to me is that we'd better not touch the logics in Filter phase. 🤔  Volumes is empty seems good enough to me. WDYT?

**Comment by kidddddddddddddddddddddd** [comment]
> A strong feeling came to me is that we'd better not touch the logics in Filter phase. 🤔 Volumes is empty seems good enough to me. WDYT?

At the beginning, I did do it this way, see [comment](https://github.com/kubernetes/kubernetes/pull/114898#discussion_r1064053975). I think we should bring @sanposhiho to this discussion.

**Comment by sanposhiho** [comment]
Hmm, so, we need to choose either of:
- to make the Skip logic simpler for ease of maintenance
  - -> only check if Volume is empty or not.
- to make the plugin fully-optimized via Skip
  - -> check further stuff to seek the possibility of returning Skip.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go
Comment (kidddddddddddddddddddddd) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: Ah right, there is a check for `pl. enableReadWriteOncePod` above😂.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:162]: Make it return true if we need to check.

```suggestion
func needsRestrictionsCheck(v *v1.Volume) bool {
```

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:168]:
```suggestion
	needsCheck := false
```

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:88]: Add a test case for a pod with a volume that doesn't have restrictions. Just in `TestGCEDiskConflicts` should be enough.

**Review by alculquicondor — COMMENTED** [review]
/approve

**Comment by alculquicondor** [comment]
I'm ok with the current implementation. The changes in Filter are trivial.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go
Comment (kidddddddddddddddddddddd) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:162]: We discussed about using pointer or struct previously in https://github.com/kubernetes/kubernetes/pull/114898#discussion_r1137295540

**Pull request merged**
